File: lintcode/easy/845_greatest_common_divisor.py
'''
Link: https://www.lintcode.com/problem/greatest-common-divisor/description
'''

# Based on the solution in Jiuzhang.com. Euclid's algorithm.
class Solution:
    """
    @param a: the given number
    @param b: another number
    @return: the greatest common divisor of two numbers
    """
    def gcd(self, a, b):
        # write your code here
        big, small = a, b
        if big < small:
            big, small = b, a
        if small != 0:
            return self.gcd(small, big % small)
        else:
            return big

File: lintcode/easy/845_greatest_common_divisor.java
/*
My own solution based on the hint given in Jiuzhang.com.
*/
public class Solution {
    /**
     * @param a: the given number
     * @param b: another number
     * @return: the greatest common divisor of two numbers
     */
    public int gcd(int a, int b) {
        // write your code here
        int max = a, min = b;
        if (max < min){
            max = b; 
            min = a;
        }
        if (min != 0){
            return gcd(min, max % min);
        }
        else{
            return max;
        }
    }
}
